src/common/ad_tx.c

```diff
@@ -0,0 +1,483 @@
+	} while (0)
+
+#define AD_TX_ACT_DEL(it)	d_list_del(it)
+
+/** allocate umem_act_item, if success the it_link and it_act.ac_opc will be init-ed */
+#define D_ALLOC_ACT(it, opc, size)							\
+	do {										\
+		if (opc == UMEM_ACT_COPY)						\
+			D_ALLOC(it, offsetof(struct umem_act_item,			\
+					     it_act.ac_copy.payload[size]));		\
+		else if (opc == UMEM_ACT_CSUM)						\
+			D_ALLOC(it, offsetof(struct umem_act_item,			\
+					     it_act.ac_csum.csum[size]));		\
+		else									\
+			D_ALLOC_PTR(it);						\
+		if (likely(it != NULL)) {						\
+			D_INIT_LIST_HEAD(&it->it_link);					\
+			it->it_act.ac_opc = opc;					\
+		}									\
+	} while (0)
+
+/** start a ad-hoc memory transaction */
+int
+ad_tx_begin(struct ad_blob *blob, struct ad_tx *tx)
+{
+	static __thread uint64_t	tx_id;
+	int				rc = 0;
+
+	tx->tx_blob = blob;
+	tx->tx_id = ++tx_id;
+	D_INIT_LIST_HEAD(&tx->tx_redo);
+	D_INIT_LIST_HEAD(&tx->tx_undo);
+
+	return rc;
+}
+
+static int
+ad_act_item_replay(struct umem_action *act)
+{
+	int rc = 0;
+
+	switch (act->ac_opc) {
+	case UMEM_ACT_NOOP:
+		break;
+	case UMEM_ACT_COPY:
+		rc = ad_tx_copy(NULL, ad_addr2ptr(act->ac_copy.addr), act->ac_copy.size,
+				act->ac_copy.payload, false);
+		break;
+	case UMEM_ACT_COPY_PTR:
+		rc = ad_tx_copy(NULL, ad_addr2ptr(act->ac_copy_ptr.addr), act->ac_copy_ptr.size,
+				ad_addr2ptr(act->ac_copy_ptr.ptr), false);
+		break;
+	case UMEM_ACT_ASSIGN:
+		rc = ad_tx_assign(NULL, ad_addr2ptr(act->ac_assign.addr), act->ac_assign.size,
+				  act->ac_assign.val);
+		break;
+	case UMEM_ACT_MOVE:
+		rc = ad_tx_move(NULL, ad_addr2ptr(act->ac_move.dst), ad_addr2ptr(act->ac_move.src),
+				act->ac_move.size);
+		break;
+	case UMEM_ACT_SET:
+		rc = ad_tx_set(NULL, ad_addr2ptr(act->ac_set.addr), act->ac_set.val,
+			       act->ac_set.size, false);
+		break;
+	case UMEM_ACT_SET_BITS:
+		rc = ad_tx_setbit(NULL, ad_addr2ptr(act->ac_op_bits.addr), act->ac_op_bits.pos,
+				  act->ac_op_bits.num);
+		break;
+	case UMEM_ACT_CLR_BITS:
+		rc = ad_tx_clrbit(NULL, ad_addr2ptr(act->ac_op_bits.addr), act->ac_op_bits.pos,
+				  act->ac_op_bits.num);
+		break;
+	case UMEM_ACT_CSUM:
+		break;
+	default:
+		rc = -DER_INVAL;
+		D_ERROR("bad ac_opc %d\n", act->ac_opc);
+		break;
+	}
+
+	if (rc)
+		D_ERROR("Failed to replay act->ac_opc %d, "DF_RC"\n", act->ac_opc, DP_RC(rc));
+
+	return rc;
+}
+
+static int
+ad_tx_act_replay(d_list_t *list)
+{
+	struct umem_act_item	*it;
+	int			 rc = 0;
+
+	d_list_for_each_entry(it, list, it_link) {
+		rc = ad_act_item_replay(&it->it_act);
+		if (rc)
+			break;
+	}
+
+	return rc;
+}
+
+static void
+ad_tx_act_cleanup(d_list_t *list)
+{
+	struct umem_act_item	*it, *next;
+
+	d_list_for_each_entry_safe(it, next, list, it_link) {
+		d_list_del(&it->it_link);
+		D_FREE(it);
+	}
+}
+
+/** complete a ad-hoc memory transaction */
+int
+ad_tx_end(struct ad_tx *tx, int err)
+{
+	int	rc;
+
+	if (err)
+		rc = ad_tx_act_replay(&tx->tx_undo);
+	else
+		rc = ad_tx_act_replay(&tx->tx_redo);
```

Comment:
here should write actions in redo list to WAL, rather than reply here. will change it later when integrate.